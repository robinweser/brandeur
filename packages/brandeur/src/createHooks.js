import { createHooks as baseCreateHooks } from '@css-hooks/react'
import { cssifyObject } from 'css-in-js-utils'

import fallbackValuePlugin from './fallbackValuePlugin'
import getFallbackVariable from './getFallbackVariable'

function processStyle(style, plugins) {
  return plugins.reduce((processed, plugin) => plugin(processed), style)
}

function resolveStyle(style, theme, keyframes) {
  if (typeof style === 'function') {
    return style({ theme, keyframes })
  }

  return style
}

function getFallbackCSS(fallbacks) {
  const rootCSS = fallbacks.reduce((css, { property, match }) => {
    // use the last value as a default matcher if no match is provided
    const actualMatch = match || values[values.length - 1]

    return (
      css +
      []
        .concat(property)
        .map((prop) => `${getFallbackVariable(prop, actualMatch)}:;`)
        .join('')
    )
  }, '')

  const supportsCSS = fallbacks.map(({ property, values = [], match }) => {
    // use the last value as a default matcher if no match is provided
    const actualMatch = match || values[values.length - 1]

    return values
      .map((value) =>
        []
          .concat(property)
          .map(
            (prop) =>
              `@supports (${prop}:${value}){:root{${getFallbackVariable(prop, actualMatch)}:${value}}}`
          )
          .join('')
      )
      .join('')
  })

  return (rootCSS ? `:root {${rootCSS}}` : '') + supportsCSS.join('')
}

function cssifyKeyframe(animationName, frames) {
  const keyframe = cssifyKeyframeRule(frames)

  return `@keyframes ${animationName}{${keyframe}}`
}

function cssifyKeyframeRule(frames) {
  return Object.keys(frames).reduce(
    (css, percentage) =>
      `${css}${percentage}{${cssifyObject(frames[percentage])}}`,
    ''
  )
}

function getKeyframesCSS(keyframes) {
  return Object.keys(keyframes).reduce(
    (css, animationName) =>
      css + cssifyKeyframe(animationName, keyframes[animationName]),
    ''
  )
}

export default function createHooks({
  hooks,
  plugins = [],
  fallbacks = [],
  keyframes = {},
  theme = {},
}) {
  if (fallbacks.length > 0) {
    plugins.push(fallbackValuePlugin(fallbacks))
  }

  const [baseCSS, fn] = baseCreateHooks(hooks)
  const fallbackCSS = getFallbackCSS(fallbacks)
  const keyframesCSS = getKeyframesCSS(keyframes)

  const staticCSS = keyframesCSS + fallbackCSS + baseCSS

  const keyframeNames = Object.keys(keyframes).reduce(
    (keyframeNames, animationName) => ({ [animationName]: animationName }),
    {}
  )

  function css(...styles) {
    return fn(
      ...styles.map((style) =>
        processStyle(resolveStyle(style, theme, keyframeNames), plugins)
      )
    )
  }

  return [staticCSS, css]
}
