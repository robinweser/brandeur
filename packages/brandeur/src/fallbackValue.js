export default function fallbackValue(property, values, match) {
  const actualMatch = match || values[values.length - 1]

  return {
    property,
    values,
    match: actualMatch,
  }
}
