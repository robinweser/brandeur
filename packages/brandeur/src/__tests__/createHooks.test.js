import beautify from 'cssbeautify'

import createHooks from '../createHooks'
import fallbackValue from '../fallbackValue'

describe('Creating hooks', () => {
  it('should support fallbacks', () => {
    const [staticCSS, css] = createHooks({
      hooks: {
        ':hover': ':hover',
      },
      fallbacks: [
        fallbackValue('position', ['-webkit-sticky', 'sticky']),
        fallbackValue(
          ['width', 'minWidth'],
          ['-webkit-min-content', 'min-content']
        ),
      ],
    })

    expect(beautify(staticCSS)).toMatchSnapshot()
    expect(
      css({
        color: 'red',
        appearance: 'none',
        position: 'sticky',
        width: 'min-content',
        minWidth: 'min-content',
        ':hover': {
          appearance: 'button',
          position: 'absolute',
          width: 'auto',
        },
      })
    ).toMatchSnapshot()
  })

  it('should support keyframes', () => {
    const [staticCSS, css] = createHooks({
      hooks: {
        ':hover': ':hover',
      },
      keyframes: {
        fadeIn: {
          from: { opacity: 0 },
          to: { opacity: 1 },
        },
      },
    })

    expect(beautify(staticCSS)).toMatchSnapshot()
    expect(
      css(({ keyframes }) => ({
        animationName: keyframes.fadeIn,
      }))
    ).toEqual({
      animationName: 'fadeIn',
    })
  })

  it('should support multiple styles', () => {
    const [_, css] = createHooks({
      hooks: {
        ':hover': ':hover',
      },
    })

    expect(css({ color: 'red' }, { fontSize: 16 })).toEqual({
      color: 'red',
      fontSize: 16,
    })
  })

  it('should support nested arrays of styles', () => {
    const [_, css] = createHooks({
      hooks: {
        ':hover': ':hover',
      },
    })

    expect(
      css(
        [
          { color: 'red', padding: 5 },
          { fontSize: 16 },
          [{ color: 'blue', backgroundColor: 'red' }, { lineHeight: 1 }],
        ],
        { padding: 10 }
      )
    ).toEqual({
      padding: 10,
      color: 'blue',
      backgroundColor: 'red',
      fontSize: 16,
      lineHeight: 1,
    })
  })
})
