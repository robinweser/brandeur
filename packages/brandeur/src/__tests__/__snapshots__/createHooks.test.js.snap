// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creating hooks should support fallbacks 1`] = `
":root {
    --position-sticky:;
    --width-min-content:;
    --minWidth-min-content:;
}

@supports (position:-webkit-sticky) {
    :root {
        --position-sticky: -webkit-sticky
    }
}

@supports (position:sticky) {
    :root {
        --position-sticky: sticky
    }
}

@supports (width:-webkit-min-content) {
    :root {
        --width-min-content: -webkit-min-content
    }
}

@supports (min-width:-webkit-min-content) {
    :root {
        --minWidth-min-content: -webkit-min-content
    }
}

@supports (width:min-content) {
    :root {
        --width-min-content: min-content
    }
}

@supports (min-width:min-content) {
    :root {
        --minWidth-min-content: min-content
    }
}

* {
    --zh6im6-0: initial;
    --zh6im6-1: ;
}

:hover {
    --zh6im6-0: ;--zh6im6-1:initial;
}"
`;

exports[`Creating hooks should support fallbacks 2`] = `
{
  "appearance": "var(--zh6im6-1, button) var(--zh6im6-0, none)",
  "color": "red",
  "minWidth": "var(--minWidth-min-content)",
  "position": "var(--zh6im6-1, absolute) var(--zh6im6-0, var(--position-sticky))",
  "width": "var(--zh6im6-1, auto) var(--zh6im6-0, var(--width-min-content))",
}
`;

exports[`Creating hooks should support keyframes 1`] = `
"@keyframes fadeIn {
    from {
        opacity: 0
    }

    to {
        opacity: 1
    }
}

* {
    --zh6im6-0: initial;
    --zh6im6-1: ;
}

:hover {
    --zh6im6-0: ;--zh6im6-1:initial;
}"
`;
